### PR TITLE
(3) As a user I should be able to create a new target so I can find people with similar interests in a specific area

### PR DESCRIPTION
**Pull request by roxuy**
[Trello card](https://trello.com/c/tXtxlI7t/3-3-as-a-user-i-should-be-able-to-create-a-new-target-so-i-can-find-people-with-similar-interests-in-a-specific-area)

### CODE REVIEW

File: app/controllers/api/v1/targets_controller.rb [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+      end
+
+      def search(radius, latitude, longitude, topic_id, current_user)
+        Target.within(
+          radius,
+          origin: [
+            latitude,
+            longitude
+          ]
+        ).where('topic_id = ? AND user_id != ?', topic_id, current_user)
```
Comment (enzofab91) [inline app/controllers/api/v1/targets_controller.rb:39]:
```suggestion
        ).where(topic_id: topic_id).where.not(user_id: current_user)
```

File: app/models/conversation.rb [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+#
+class Conversation < ApplicationRecord
+  belongs_to :user1, class_name: 'User', foreign_key: 'user1'
+  belongs_to :user2, class_name: 'User', foreign_key: 'user2'
+  belongs_to :topic
+  validates :user1, presence: true
+  validates :user2, presence: true
+  validates :topic_id, presence: true
+  validate :users_must_be_different
+  def users_must_be_different
```
Comment (enzofab91) [inline app/models/conversation.rb:28]: Add an empty line here.

File: spec/request/api/v1/target/targets_spec.rb [hunk 3]
```diff
@@ -0,0 +1,89 @@
+require 'rails_helper'
+
+RSpec.describe 'Target /api/v1/targets', type: :request do
+  let(:user) { create(:user, confirmed_at: Time.zone.today) }
+  let!(:user2) { create(:user, confirmed_at: Time.zone.today) }
+  let(:topic) { create(:topic) }
+  let!(:target) do
+    create :target, user: user2, topic_id: topic.id,
+                    latitude: -32.528879, longitude: -55.769835, radius: 400
```
Comment (enzofab91) [inline spec/request/api/v1/target/targets_spec.rb:9]:
```suggestion
    create(:target, user: user2, topic_id: topic.id,
                    latitude: -32.528879, longitude: -55.769835, radius: 400)
```

File: spec/request/api/v1/target/targets_spec.rb [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+
+  it 'should creates the target' do
+    expect {
+      subject
+    }.to change(Target, :count).by(1)
+  end
+
+  it 'should returns the created target' do
+    subject
+    expect(json['target']['title']).to eq('target title')
```
Comment (enzofab91) [inline spec/request/api/v1/target/targets_spec.rb:43]: You can use `let` to define variables in rspec.

See https://relishapp.com/rspec/rspec-core/v/2-11/docs/helper-methods/let-and-let

**Review by enzofab91 — COMMENTED** [review]
Good work!

Left some comments. Check failing tests 🙏

**Review by enzofab91 — APPROVED** [review]
Good work!

File: app/controllers/api/v1/targets_controller.rb [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+                                 target.topic_id, target.user)
+
+        conversations = create_conversations(targets, target)
+
+        matched_users = targets.map(&:user).uniq
+
+        { target: target, conversations: conversations, matched_users: matched_users }
+      end
+
+      def search_targets(radius, latitude, longitude, topic_id, current_user)
```
Comment (sebastiancaraballo) [inline app/controllers/api/v1/targets_controller.rb:33]: All this logic should be extracted into a service class. Maybe something like: `app/services/target_matcher_service.rb`

File: app/models/conversation.rb [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+# Foreign Keys
+#
+#  fk_rails_...  (topic_id => topics.id)
+#
+class Conversation < ApplicationRecord
+  belongs_to :user1, class_name: 'User', foreign_key: 'user1', inverse_of: 'conversations1'
+  belongs_to :user2, class_name: 'User', foreign_key: 'user2', inverse_of: 'conversations2'
+  belongs_to :topic
+  validates :user1, presence: true
+  validates :user2, presence: true
```
Comment (sebastiancaraballo) [inline app/models/conversation.rb:25]:
```suggestion
  validates :user1, :user2, presence: true
```

File: app/models/conversation.rb [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+#  index_conversations_on_topic_id  (topic_id)
+#
+# Foreign Keys
+#
+#  fk_rails_...  (topic_id => topics.id)
+#
+class Conversation < ApplicationRecord
+  belongs_to :user1, class_name: 'User', foreign_key: 'user1', inverse_of: 'conversations1'
+  belongs_to :user2, class_name: 'User', foreign_key: 'user2', inverse_of: 'conversations2'
+  belongs_to :topic
```
Comment (sebastiancaraballo) [inline app/models/conversation.rb:23]:
```suggestion
  belongs_to :topic
  
```

File: app/models/target.rb [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+#
+#  fk_rails_...  (topic_id => topics.id)
+#  fk_rails_...  (user_id => users.id)
+#
+class Target < ApplicationRecord
+  acts_as_mappable default_units: :kms,
+                   default_formula: :sphere,
+                   distance_field_name: :radius,
+                   lat_column_name: :latitude,
+                   lng_column_name: :longitude
```
Comment (sebastiancaraballo) [inline app/models/target.rb:32]:
```suggestion
                   lng_column_name: :longitude
                   
```

File: app/models/target.rb [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+  belongs_to :topic
+  belongs_to :user
```
Comment (sebastiancaraballo) [inline app/models/target.rb:34]:
```suggestion
  belongs_to :user
  
```

File: app/services/target_matcher_service.rb [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+      origin: [
+        latitude,
+        longitude
+      ]
+    ).where(topic_id: topic_id).where.not(user_id: current_user)
+  end
+
+  def create_conversations(targets, target)
+    targets.map do |t|
+      Conversation.create(user1: target.user, user2: User.find(t.user.id), topic: target.topic)
```
Comment (sebastiancaraballo) [inline app/services/target_matcher_service.rb:28]:
```suggestion
      Conversation.create!(user1: target.user, user2: User.find(t.user.id), topic: target.topic)
```

File: db/migrate/20210827172413_create_targets.rb [hunk 11]
```diff
@@ -0,0 +1,16 @@
+class CreateTargets < ActiveRecord::Migration[6.1]
+  def change
+    create_table :targets do |t|
+      t.string :title, null: false
+      t.references :topic, null: false, foreign_key: true
+      t.references :user, null: false, foreign_key: true
+      t.decimal :latitude, null: false, precision: 10, scale: 6
+      t.decimal :longitude, null: false, precision: 10, scale: 6
+      t.decimal :radius, null: false, precision: 7, scale: 2
```
Comment (sebastiancaraballo) [inline db/migrate/20210827172413_create_targets.rb:9]: I recommend using `float`

File: spec/models/conversation_spec.rb [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+# Foreign Keys
+#
+#  fk_rails_...  (topic_id => topics.id)
+#
+require 'rails_helper'
+
+RSpec.describe Conversation, type: :model do
+  let(:user1) { create :user }
+  let(:user2) { create :user }
+  let(:topic) { create :topic }
```
Comment (sebastiancaraballo) [inline spec/models/conversation_spec.rb:25]:
```suggestion
  let(:topic) { create :topic }

```

**Review by sebastiancaraballo — APPROVED** [review]

**Pull request merged**